src/Network/HTTP/Affjax.purs

```diff
@@ -117,6 +123,41 @@ delete u = affjax $ defaultRequest { method = DELETE, url = u }
 delete_ :: forall e. URL -> Affjax e Unit
 delete_ = delete
 
+-- | Retry a request with exponential backoff, timing out optionally after a specified number of milliseconds.
+retry :: forall e a b. (Requestable a) => Maybe Int -> (AffjaxRequest a -> Affjax (avar :: AVAR | e) b) -> (AffjaxRequest a -> Affjax (avar :: AVAR | e) b)
+retry milliseconds run req = do
+  failureVar <- makeVar
+  let loop = go failureVar
+  case milliseconds of
+    Nothing -> loop 1
+    Just milliseconds -> do
+      respVar <- makeVar
+      loopHandle <- forkAff $ loop 1 >>= putVar respVar <<< Just
+      timeoutHandle <-
+        forkAff <<< later' milliseconds $ do
+          putVar respVar Nothing
+          loopHandle `cancel` error "Cancel"
+      result <- takeVar respVar
+      case result of
+        Nothing ->
+          takeVar failureVar
+        Just resp -> pure resp
+  where
+    assert200 resp =
+      case resp.status of
+        StatusCode 200 -> Right resp
+        _ -> Left resp
+
+    go failureVar n = do
+      result <- run req
+      case assert200 result of
+        Right b -> pure b
+        Left resp -> do
+          putVar failureVar resp
+          -- TODO: is this too steep?
+          let delay = round $ toNumber 1000 * pow (toNumber 2) (toNumber (n - 1))
```

Comment:
Maybe a tad too steep, you may also want to cap this since after a certain point, there's no additional benefit in delaying.

The main point of backoff is to give an overloaded server a chance to recover; secondary point is to avoid choking up network bandwidth on something that's going to continuously fail until the user re-acquires Internet connection.

For web apps like SlamData which are designed to be left open in browsers as a user travels, we'll probably specify no time out, so that upper cap on retry time will determine latency when the user re-acquires connectivity (with that in mind, I'd say on the order of 10 - 20 seconds is an appropriate upper bound, maybe as high as 60 seconds).